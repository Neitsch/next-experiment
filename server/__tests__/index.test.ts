describe("Server", () => {
  it("Starts", () => {
    const reqHandler = jest.fn();
    let serverLauncher: () => void;
    let errorFunc: (v?: any) => void;
    let setPort;
    const consoleErrorSave = console.error;
    console.error = jest.fn();
    const promise = {
      catch: jest.fn(),
    };
    const nonce = "abc";
    jest.doMock("uuid", () => ({
      v4: jest.fn().mockReturnValue(nonce),
    }));
    jest.doMock("../../lib/routes", () => ({
      getRequestHandler: () => reqHandler,
    }));
    jest.doMock("next", () => () => ({
      prepare: () => ({
        then: (fn: any) => {
          serverLauncher = fn;
          return promise;
        },
      }),
    }));
    const innerGql = jest.fn();
    const getFn = jest.fn();
    const gqlFn = jest.fn().mockReturnValue(innerGql);
    const useFn = jest.fn();
    jest.doMock("body-parser", () => ({ json: jest.fn() }));
    jest.doMock("compression", () => jest.fn());
    jest.doMock("cookie-parser", () => jest.fn());
    jest.doMock("cors", () => jest.fn());
    jest.doMock("express", () => () => ({
      get: getFn,
      listen: jest.fn().mockImplementation((port, fn2) => {
        setPort = port;
        errorFunc = fn2;
      }),
      use: useFn,
    }));
    jest.doMock("express-jwt", () => jest.fn());
    jest.doMock("jwks-rsa", () => ({
      expressJwtSecret: jest.fn(),
    }));
    jest.doMock("apollo-server-express", () => ({
      graphiqlExpress: jest.fn(),
      graphqlExpress: gqlFn,
    }));
    jest.doMock("graphql-validation-complexity");
    require("../index");
    serverLauncher!();
    expect(reqHandler).toHaveBeenCalledTimes(0);
    const url = "http://www.abc.com/123";
    const urlCall = { url };
    const otherArg = {};
    getFn.mock.calls[0][1](urlCall, otherArg);
    expect(reqHandler).toHaveBeenCalledTimes(1);
    expect(reqHandler).toHaveBeenCalledWith(urlCall, otherArg);
    expect(setPort).toEqual(3000);
    expect(() => {
      errorFunc!(new Error(":)"));
    }).toThrow();
    expect(() => {
      errorFunc!();
    }).not.toThrow();
    const reqIn = { user: { sub: "SUB" } };
    useFn.mock.calls[5][3](reqIn, otherArg);
    expect(useFn.mock.calls[5][3]).toEqual(innerGql);
    expect(
      gqlFn.mock.calls[0][0]({ user: { sub: "MySub" } }),
    ).toMatchSnapshot();
    expect(gqlFn.mock.calls[0][0](null)).toMatchSnapshot();
    expect(() => {
      promise.catch.mock.calls[0][0]("Test");
    }).not.toThrow();
    const resFn = jest.fn();
    useFn.mock.calls[7][0](new Error("test"), null, { json: resFn });
    expect(resFn.mock.calls).toMatchSnapshot();
    const cb = jest.fn();
    const res = { locals: {} };
    useFn.mock.calls[4][0]({}, res, cb);
    expect(cb).toHaveBeenCalled();
    expect(res.locals).toHaveProperty("nonce");
    expect(console.error).toHaveBeenCalledWith("Test");
    console.error = consoleErrorSave;
  });
});
