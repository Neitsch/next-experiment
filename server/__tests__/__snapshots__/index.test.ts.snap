// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server Starts 1`] = `
Object {
  "cacheControl": true,
  "context": Object {
    "userSub": "MySub",
  },
  "formatError": [Function],
  "rootValue": "MySub",
  "schema": GraphQLSchema {
    "__allowedLegacyNames": undefined,
    "_directives": Array [
      GraphQLDirective {
        "args": Array [
          Object {
            "astNode": undefined,
            "defaultValue": undefined,
            "description": "Included when true.",
            "name": "if",
            "type": "Boolean!",
          },
        ],
        "astNode": undefined,
        "description": "Directs the executor to include this field or fragment only when the \`if\` argument is true.",
        "locations": Array [
          "FIELD",
          "FRAGMENT_SPREAD",
          "INLINE_FRAGMENT",
        ],
        "name": "include",
      },
      GraphQLDirective {
        "args": Array [
          Object {
            "astNode": undefined,
            "defaultValue": undefined,
            "description": "Skipped when true.",
            "name": "if",
            "type": "Boolean!",
          },
        ],
        "astNode": undefined,
        "description": "Directs the executor to skip this field or fragment when the \`if\` argument is true.",
        "locations": Array [
          "FIELD",
          "FRAGMENT_SPREAD",
          "INLINE_FRAGMENT",
        ],
        "name": "skip",
      },
      GraphQLDirective {
        "args": Array [
          Object {
            "astNode": undefined,
            "defaultValue": "No longer supported",
            "description": "Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted in [Markdown](https://daringfireball.net/projects/markdown/).",
            "name": "reason",
            "type": "String",
          },
        ],
        "astNode": undefined,
        "description": "Marks an element of a GraphQL schema as no longer supported.",
        "locations": Array [
          "FIELD_DEFINITION",
          "ENUM_VALUE",
        ],
        "name": "deprecated",
      },
    ],
    "_implementations": Object {},
    "_mutationType": undefined,
    "_queryType": "Query",
    "_subscriptionType": undefined,
    "_typeMap": Object {
      "Boolean": "Boolean",
      "Query": "Query",
      "String": "String",
      "__Directive": "__Directive",
      "__DirectiveLocation": "__DirectiveLocation",
      "__EnumValue": "__EnumValue",
      "__Field": "__Field",
      "__InputValue": "__InputValue",
      "__Schema": "__Schema",
      "__Type": "__Type",
      "__TypeKind": "__TypeKind",
    },
    "astNode": undefined,
  },
  "tracing": true,
  "validationRules": Array [
    undefined,
  ],
}
`;

exports[`Server Starts 2`] = `
Array [
  Array [
    Object {
      "errors": Array [
        Object {
          "message": "test",
          "name": "Error",
        },
      ],
    },
  ],
]
`;
